### PR TITLE
Issues 1039 979

### PR DESCRIPTION
**Pull request by anfernee**
- Implement Destroy task for HostSystem
-  nit PortKeys in DistributedVirtualPortgroup

### CODE REVIEW

File: object/host_system.go [hunk 1]
```diff
@@ -151,3 +151,16 @@ func (h HostSystem) ExitMaintenanceMode(ctx context.Context, timeout int32) (*Ta
 
 	return NewTask(h.c, res.Returnval), nil
 }
+
+func (h HostSystem) Destroy(ctx context.Context) (*Task, error) {
```
Comment (dougm) [inline object/host_system.go]: No need to add this method, it is already defined in the embedded `Common` struct.

File: simulator/host_system.go [hunk 2]
```diff
@@ -173,6 +173,20 @@ func CreateStandaloneHost(f *Folder, spec types.HostConnectSpec) (*HostSystem, t
 	return host, nil
 }
 
+func (h *HostSystem) DestroyTask(req *types.Destroy_Task) soap.HasFault {
+	task := CreateTask(h, "destroy", func(t *Task) (types.AnyType, types.BaseMethodFault) {
+		f := Map.getEntityParent(h, "Folder").(*Folder)
```
Comment (dougm) [inline simulator/host_system.go]: We probably need to check that the host doesn't have any VMs running on it, may need to be in disconnected state, etc.  ok to leave as-is for now though, maybe just add a TODO.

File: simulator/host_system_test.go [hunk 3]
```diff
@@ -130,3 +130,38 @@ func TestNewHostSystem(t *testing.T) {
 		t.Fatal("expected hs.Summary.Runtime == &hs.Runtime; got !=")
 	}
 }
+
+func TestDestroyHostSystem(t *testing.T) {
+	ctx := context.Background()
+	m := ESX()
```
Comment (dougm) [inline simulator/host_system_test.go]: Don't think real ESX allows Destroy on the 1 HostSystem.  But this is also fine to leave as-is for now.

File: simulator/host_system.go (hunk 2 above)
Comment (anfernee) [reply simulator/host_system.go]: @dougm I see. Will make the change.

File: simulator/host_system_test.go (hunk 3 above)
Comment (anfernee) [reply simulator/host_system_test.go]: got it. gonna check why it passed..

File: simulator/host_system.go [hunk 4]
```diff
@@ -173,6 +173,26 @@ func CreateStandaloneHost(f *Folder, spec types.HostConnectSpec) (*HostSystem, t
 	return host, nil
 }
 
+func (h *HostSystem) DestroyTask(req *types.Destroy_Task) soap.HasFault {
+	task := CreateTask(h, "destroy", func(t *Task) (types.AnyType, types.BaseMethodFault) {
+		for _, vmref := range h.Vm {
+			vm := Map.Get(vmref).(*VirtualMachine)
+			vm.Runtime.ConnectionState = types.VirtualMachineConnectionStateDisconnected
```
Comment (dougm) [inline simulator/host_system.go]: Don't want to change ConnectionState here "When the API is provided directly by a server product, such as ESX Server, then the disconnected state is not possible", see: http://pubs.vmware.com/vsphere-6-5/index.jsp#com.vmware.wssdk.apiref.doc/vim.VirtualMachine.ConnectionState.html

Also note that with the new locking mechanism, the HostSystem (`h`) will be locked here, but VirtualMachine will not.  So you'd want to use `Map.WithLock(vm, func() {...})`

I think the task should return an error such as `ResourceInUse` if `len(h.Vm) != 0`, like what happens against real VC.
A host can't be destroyed unless VMs have been migrated elsewhere (or VMs have been deleted).  There's probably other checks that real VC makes, but again I think it's fine to leave those checks out for now.

Comment (anfernee) [reply simulator/host_system.go]: AH. I see. Thanks for pointing out the correct behavior.

File: simulator/host_system.go [hunk 5]
```diff
@@ -173,6 +173,31 @@ func CreateStandaloneHost(f *Folder, spec types.HostConnectSpec) (*HostSystem, t
 	return host, nil
 }
 
+func (h *HostSystem) DestroyTask(req *types.Destroy_Task) soap.HasFault {
+	if len(h.Vm) > 0 {
```
Comment (dougm) [inline simulator/host_system.go]: Don't do this here, real VC doesn't fail the creation of tasks, only the running of tasks (which you have right below).

**Comment by anfernee** [comment]
cc @dougm

**Pull request merged**